gateway/upstreamer/push/options.go

```diff
@@ -10,20 +14,35 @@ type upstreamConfig struct {
 	overrideEndpointAddress     string
 	exposePrivateAPIs           bool
 	eventsAPIs                  map[string]string
+	feedbackLoopSamples         int
+	lock                        sync.Mutex
 	requiredServices            []string
 	serviceTimeout              time.Duration
 	serviceTimeoutCheckInterval time.Duration
-	loadBasedBalancerFunc       LoadBasedBalancerFunc
-	loadBasedBalancerThreshold  int
+	randomizer                  Randomizer
 }
 
 func newUpstreamConfig() upstreamConfig {
 	return upstreamConfig{
 		eventsAPIs:                  map[string]string{},
+		feedbackLoopSamples:         100,
+		lock:                        sync.Mutex{},
 		serviceTimeout:              30 * time.Second,
 		serviceTimeoutCheckInterval: 5 * time.Second,
-		loadBasedBalancerThreshold:  6,
-		loadBasedBalancerFunc:       DefaulLoadBasedBalancerFunc,
+		randomizer:                  rand.New(rand.NewSource(time.Now().UnixNano())),
+	}
+}
+
+// A Randomizer reprensents an interface to randomize
+type Randomizer interface {
+	Intn(int) int
+	Shuffle(n int, swap func(i, j int))
```

Comment:
I think the struct implementing this interface should hold the lock. Right now it's asymmetrical